src/librustc_typeck/check/generator_interior.rs

```diff
@@ -0,0 +1,111 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use log;
+use rustc::hir::def_id::DefId;
+use rustc::hir::intravisit::{self, Visitor, NestedVisitorMap};
+use rustc::hir::{self, Body, Pat, PatKind, Expr};
+use rustc::middle::region::{RegionMaps, CodeExtent};
+use rustc::ty::Ty;
+use std::rc::Rc;
+use super::FnCtxt;
+use util::nodemap::FxHashSet;
+
+struct InteriorVisitor<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
+    fcx: &'a FnCtxt<'a, 'gcx, 'tcx>,
+    types: FxHashSet<Ty<'tcx>>,
+    region_maps: Rc<RegionMaps>,
+}
+
+impl<'a, 'gcx, 'tcx> InteriorVisitor<'a, 'gcx, 'tcx> {
+    fn record(&mut self, ty: Ty<'tcx>, scope: Option<CodeExtent>, expr: Option<&'tcx Expr>) {
+        use syntax_pos::DUMMY_SP;
+
+        if scope.map(|s| self.fcx.tcx.yield_in_extent(s).is_some()).unwrap_or(true) {
+            if log_enabled!(log::LogLevel::Debug) {
+                if let Some(s) = scope {
+                    let span = s.span(&self.fcx.tcx.hir).unwrap_or(DUMMY_SP);
+                    debug!("type in generator with scope = {:?}, type = {:?}, span = {:?}",
```

Comment:
for extra readability, include the `expr` in both cases here. Most of typeck has `debug` statements like that, and it's plenty readable.